src/buffer_manager.ts

```diff
@@ -425,6 +428,104 @@ export class BufferManager implements Disposable {
 
     private handleWindowChangedDebounced = debounce(this.handleWindowChanged, 100, { leading: false, trailing: true });
 
+    private async syncDocumentDirtyState(): Promise<void> {
+        const states = Array.from(this.textDocumentToBufferId.entries()).map(([doc, bufId]) => ({
+            buf: bufId,
+            modified: doc.isDirty,
+        }));
+        await this.client.lua(
+            `
+            local states = ...
+            for _, state in ipairs(states) do
+                vim.bo[state.buf].modified = state.modified
+            end
+            `,
+            [states],
+        );
+    }
+
+    private async handleSaveBuf({
+        buf,
+        bang,
+        current_name,
+        target_name,
+    }: {
+        buf: number;
+        bang: boolean;
+        current_name: string;
+        target_name: string;
+    }) {
+        // Note: workspace.save and workspace.saveAs are smart enough to handle
+        // the documents that are not a real file (e.g. untitled, output, etc.)
+        // so we can just call them directly
+
+        const document = this.getTextDocumentForBufferId(buf);
+        if (document == null) {
+            throw new Error("Received an invalid buffer id from Neovim, cannot save");
+        }
+
+        const docUri = document.uri;
+
+        if (document.isUntitled) {
+            await workspace.save(docUri);
+            return;
+        }
+
+        // If using Windows locally and developing on a Unix remote environment,
+        // the saved path can contain backslashes, causing folders to be treated as filenames.
+        const normalize = (p: string) => path.normalize(p).split(path.sep).join(path.posix.sep);
+
+        const currentPath = normalize(current_name);
+        const targetPath = normalize(target_name);
+
+        if (currentPath === targetPath) {
+            await workspace.save(docUri);
+            return;
+        }
+
+        const vimCwd = normalize(await this.main.client.call("getcwd"));
```

Comment:
Instead of asking vim for the cwd (which we currently provide through hacks), can we get the cwd from the workspace or document?

Reply:
The approach here is to compute a generic relative file path representing the current operation, always using Vim data for computation first, and then integrating with the working directory in VSCode for subsequent operations. Vim-cwd may not always be accurate or set. A straightforward example is during remote development where nvim runs locally.


1. vim-target-filepath + vim-cwd => relative-path
2. relative-path + vscode-cwd => vscode-target-filepath

Reply:
that comment + example seems worth adding as a code comment